File: src/ts/ssh-keys/publicKeyFormatter.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { SshAlgorithms, KeyPair, Rsa, ECDsa, SshDataReader } from '@microsoft/dev-tunnels-ssh';
import { KeyFormatter } from './keyFormatter';
import { KeyData } from './keyData';

/** Provides import/export of the SSH public key format. */
export class PublicKeyFormatter implements KeyFormatter {
	public async import(keyData: KeyData): Promise<KeyPair | null> {
		if (!keyData) throw new TypeError('KeyData object expected.');

		if (!keyData.keyType) {
			// Try to parse binary data without any key type prefix.
			try {
				const reader = new SshDataReader(keyData.data);
				keyData.keyType = reader.readString('ascii');
			} catch (e) {
				return null;
			}
		}

		let keyPair: KeyPair | null = null;

		if (keyData.keyType === Rsa.keyAlgorithmName) {
			keyPair = SshAlgorithms.publicKey.rsaWithSha512!.createKeyPair();
		} else if (keyData.keyType === ECDsa.ecdsaSha2Nistp256) {
			keyPair = SshAlgorithms.publicKey.ecdsaSha2Nistp256!.createKeyPair();
		} else if (keyData.keyType === ECDsa.ecdsaSha2Nistp384) {
			keyPair = SshAlgorithms.publicKey.ecdsaSha2Nistp384!.createKeyPair();
		} else if (keyData.keyType === ECDsa.ecdsaSha2Nistp521) {
			keyPair = SshAlgorithms.publicKey.ecdsaSha2Nistp521!.createKeyPair();
		}

		if (keyPair) {
			await keyPair.setPublicKeyBytes(keyData.data);

			const comment = keyData.headers.get('Comment');
			keyPair.comment = comment ?? null;
		}

		return keyPair;
	}

	public async export(keyPair: KeyPair, includePrivate: boolean): Promise<KeyData> {
		if (!keyPair) throw new TypeError('KeyPair object expected.');

		if (includePrivate) {
			throw new Error('SSH public key formatter does not support private keys.');
		}

		if (!keyPair.hasPublicKey) {
			throw new Error('KeyPair object does not include a public key.');
		}

		const keyData = new KeyData();
		keyData.keyType = keyPair.keyAlgorithmName;
		keyData.data = (await keyPair.getPublicKeyBytes())!;

		if (keyPair.comment) {
			keyData.headers.set('Comment', keyPair.comment);
		}

		return keyData;
	}

	public async decrypt(keyData: KeyData, passphrase: string | null): Promise<KeyData | null> {
		return keyData;
	}

	public encrypt(keyData: KeyData, passphrase: string): Promise<KeyData> {
		throw new Error('SSH public key format does not support encryption.');
	}
}

File: src/cs/Ssh.Keys/PublicKeyFormatter.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Text;
using Microsoft.DevTunnels.Ssh.Algorithms;
using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Ssh.Keys;

/// <summary>
/// Provides import/export of the <see cref="KeyFormat.Ssh" /> public key format.
/// </summary>
public class PublicKeyFormatter : IKeyFormatter
{
	/// <inheritdoc/>
	public IKeyPair? Import(KeyData keyData)
	{
		if (keyData == null) throw new ArgumentNullException(nameof(keyData));

		if (string.IsNullOrEmpty(keyData.KeyType))
		{
			// Try to parse binary data without any key type prefix.
			try
			{
				var reader = new SshDataReader(keyData.Data);
				keyData.KeyType = reader.ReadString(Encoding.ASCII);
			}
			catch (Exception)
			{
				return null;
			}
		}

		if (keyData.KeyType == Rsa.KeyAlgorithmName)
		{
			var keyPair = new Rsa.KeyPair();
			keyPair.SetPublicKeyBytes(keyData.Data);
			keyPair.Comment = keyData.Headers.TryGetValue("Comment", out var comment)
				? comment : null;
			return keyPair;
		}
		else if (keyData.KeyType == ECDsa.ECDsaSha2Nistp256 ||
			keyData.KeyType == ECDsa.ECDsaSha2Nistp384 ||
			keyData.KeyType == ECDsa.ECDsaSha2Nistp521)
		{
			var keyPair = new ECDsa.KeyPair();
			keyPair.SetPublicKeyBytes(keyData.Data);
			keyPair.Comment = keyData.Headers.TryGetValue("Comment", out var comment)
				? comment : null;
			return keyPair;
		}

		return null;
	}

	/// <inheritdoc/>
	public KeyData Export(IKeyPair keyPair, bool includePrivate)
	{
		if (keyPair == null) throw new ArgumentNullException(nameof(keyPair));

		if (includePrivate)
		{
			throw new NotSupportedException(
				"SSH public key formatter does not support private keys.");
		}

		var keyData = new KeyData();
		keyData.KeyType = keyPair.KeyAlgorithmName;
		keyData.Data = keyPair.GetPublicKeyBytes().ToArray();

		if (!string.IsNullOrEmpty(keyPair.Comment))
		{
			keyData.Headers["Comment"] = keyPair.Comment!;
		}

		return keyData;
	}

	/// <inheritdoc/>
	public KeyData? Decrypt(KeyData keyData, string? passphrase) => keyData;

	/// <inheritdoc/>
	public KeyData Encrypt(KeyData keyData, string? passphrase) =>
		throw new NotSupportedException("SSH public key format does not support encryption.");
}
